File: backend/app/api/findings.py
"""
Assurance Findings REST API Endpoints
"""
import json
from typing import Optional, List
from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.orm import Session
from backend.app.database.database import get_db
from backend.app.database.repository import Repository

router = APIRouter(prefix="/findings", tags=["Findings"])

@router.get("")
def list_findings(
    asset_id: Optional[str] = Query(None),
    category: Optional[str] = Query(None),
    severity: Optional[str] = Query(None),
    db: Session = Depends(get_db)
):
    """List and filter assurance findings."""
    repo = Repository(db)
    findings = repo.list_findings(asset_id=asset_id, category=category, severity=severity)
    
    return [
        {
            "id": f.id,
            "finding_id": f.finding_id,
            "asset_type": f.asset_type,
            "asset_id": f.asset_id,
            "category": f.category,
            "severity": f.severity,
            "confidence": f.confidence,
            "risk_score": f.risk_score,
            "title": f.title,
            "description": f.description,
            "recommendation": f.recommendation,
            "detection_method": f.detection_method,
            "created_at": f.created_at.isoformat() if f.created_at else None
        }
        for f in findings
    ]

@router.get("/{finding_id}")
def get_finding_detail(finding_id: str, db: Session = Depends(get_db)):
    """Retrieve deep evidence, affected samples, limitations, and recommendation for a finding."""
    repo = Repository(db)
    f = repo.get_finding(finding_id)
    if not f:
        raise HTTPException(status_code=404, detail="Finding not found")

    return {
        "finding_id": f.finding_id,
        "asset_type": f.asset_type,
        "asset_id": f.asset_id,
        "category": f.category,
        "severity": f.severity,
        "confidence": f.confidence,
        "risk_score": f.risk_score,
        "title": f.title,
        "description": f.description,
        "evidence": json.loads(f.evidence) if f.evidence else [],
        "affected_samples": json.loads(f.affected_samples) if f.affected_samples else [],
        "detection_method": f.detection_method,
        "limitations": json.loads(f.limitations) if f.limitations else [],
        "recommendation": f.recommendation,
        "created_at": f.created_at.isoformat() if f.created_at else None
    }

File: frontend/js/findings.js
/**
 * VisionTrust Assurance - Findings Page Logic
 */
import { API, showToast, getRiskBadge } from './api.js';

let allFindings = [];

document.addEventListener('DOMContentLoaded', async () => {
  await loadFindings();

  // Filter listeners
  const filterSev = document.getElementById('filter-severity');
  const filterCat = document.getElementById('filter-category');
  const filterAsset = document.getElementById('filter-asset');

  if (filterSev) filterSev.addEventListener('change', () => applyFilters());
  if (filterCat) filterCat.addEventListener('change', () => applyFilters());
  if (filterAsset) filterAsset.addEventListener('input', () => applyFilters());

  // Drawer close listeners
  const btnClose = document.getElementById('btn-close-drawer');
  const overlay = document.getElementById('drawer-overlay');
  if (btnClose) btnClose.addEventListener('click', () => closeDrawer());
  if (overlay) overlay.addEventListener('click', () => closeDrawer());
});

async function loadFindings() {
  try {
    const findings = await API.listFindings();
    allFindings = findings;
    applyFilters();
  } catch (err) {
    showToast(`Failed to load findings: ${err.message}`, 'error');
  }
}

function applyFilters() {
  const sev = document.getElementById('filter-severity').value;
  const cat = document.getElementById('filter-category').value;
  const asset = document.getElementById('filter-asset').value.trim().toLowerCase();

  let filtered = allFindings;

  if (sev) {
    filtered = filtered.filter(f => f.severity === sev);
  }
  if (cat) {
    filtered = filtered.filter(f => f.category === cat);
  }
  if (asset) {
    filtered = filtered.filter(f => (f.asset_id || '').toLowerCase().includes(asset));
  }

  renderFindingsTable(filtered);
}

function renderFindingsTable(findings) {
  const tbody = document.getElementById('findings-tbody');
  if (!tbody) return;

  if (!findings || findings.length === 0) {
    tbody.innerHTML = `<tr><td colspan="9" style="text-align: center; color: var(--text-dim); padding: 2rem;">No findings matching current filters.</td></tr>`;
    return;
  }

  tbody.innerHTML = findings.map(f => `
    <tr>
      <td><code style="color: var(--accent-cyan);">${f.finding_id}</code></td>
      <td><span style="font-size: 0.78rem; color: var(--text-dim);">${f.asset_type}:</span> ${f.asset_id}</td>
      <td><span class="badge badge-info">${f.category}</span></td>
      <td>${getRiskBadge(f.severity)}</td>
      <td>${(f.confidence * 100).toFixed(0)}%</td>
      <td><strong>${f.risk_score.toFixed(1)}</strong></td>
      <td>${f.title}</td>
      <td>${getRiskBadge(f.recommendation)}</td>
      <td>
        <button class="btn btn-secondary btn-sm btn-open-finding" data-id="${f.finding_id}">
          Evidence
        </button>
      </td>
    </tr>
  `).join('');

  document.querySelectorAll('.btn-open-finding').forEach(btn => {
    btn.addEventListener('click', async (e) => {
      const id = e.target.getAttribute('data-id');
      await openFindingDrawer(id);
    });
  });
}

async function openFindingDrawer(findingId) {
  try {
    const f = await API.getFinding(findingId);
    
    document.getElementById('drawer-title').textContent = f.title;
    document.getElementById('drawer-subtitle').textContent = `${f.finding_id} | ${f.asset_type}: ${f.asset_id} | ${f.category}`;
    document.getElementById('drawer-desc').textContent = f.description;
    document.getElementById('drawer-method').textContent = f.detection_method;

    // Evidence
    const evBlock = document.getElementById('drawer-evidence');
    if (evBlock) {
      evBlock.textContent = f.evidence && f.evidence.length > 0 ? f.evidence.join('\n\n') : 'No additional telemetry recorded.';
    }

    // Samples
    const samplesBlock = document.getElementById('drawer-samples');
    if (samplesBlock) {
      samplesBlock.textContent = f.affected_samples && f.affected_samples.length > 0 ? f.affected_samples.join(', ') : 'Asset-level scope (no isolated sample list).';
    }

    // Limitations
    const limBlock = document.getElementById('drawer-limitations');
    if (limBlock) {
      limBlock.textContent = f.limitations && f.limitations.length > 0 ? f.limitations.join('\n') : 'Standard algorithmic constraints apply.';
    }

    // Recommendation
    document.getElementById('drawer-rec').innerHTML = `${getRiskBadge(f.recommendation)} &nbsp; Risk Score: ${f.risk_score.toFixed(1)}`;

    document.getElementById('evidence-drawer').classList.add('open');
    document.getElementById('drawer-overlay').classList.add('open');
  } catch (err) {
    showToast(`Error retrieving finding details: ${err.message}`, 'error');
  }
}

function closeDrawer() {
  document.getElementById('evidence-drawer').classList.remove('open');
  document.getElementById('drawer-overlay').classList.remove('open');
}
